### DISCUSSION TITLE
Performance compared to raw filesystem

### DISCUSSION THREAD

**Discussion by JaffaKetchup** [isar/isar#965] [Help] [answered]
Hello all,

I am an owner and maintainer of a [plugin package](https://github.com/JaffaKetchup/flutter_map_tile_caching) that provides advanced caching capabilities for the most popular non-commercially owned [mapping library](https://github.com/fleaflet/flutter_map) in Flutter. Performance is of utmost importance for the package, so this package is looking good.

I would like to know whether anyone has measured whether it is faster to store images directly on the filesystem, or as blobs in the database? Currently, my package creates its own structure directly on the filesystem, but I would like to try to move away from this. Of course, I realise that this database will be faster for things like sorting, but in just pure read/write performance, would anyone have a good idea?

Please see https://fmtc.jaffaketchup.dev/roadmap#migrate-to-database-format-internally if you would like a little more context or have any interest in my aims!

Many thanks :)

**Comment by simc** [comment, marked as answer]
Storing big blobs in a database is generally not a good idea. Likely the fastest approach is storing metadata in Isar and the actual blobs in the file system.

**Reply by JaffaKetchup** [reply]
Hey @simc, thanks for the fast response!

Can you explain a little further as to why storing blobs in a database isn't a great idea? For context, a standard map tile image is 256x256 pixels, perhaps an average of - let's just say - only 30-40KB. The biggest total amount might be 200MB (of course, this would vary, but the vast majority would be under this).

I understand the risk of the one file becoming very large, but this should be somewhat mitigated in my case, as I plan to have multiple database files (one for every 'store'). Perhaps this isn't the reason?

**Reply by simc** [reply]
Around 200MB is totally fine and performance will be much better than reading from the file system.

Anything larger than 1-2GB becomes an issue because Isar user virtual memory and there's a limit on some older devices

**Reply by JaffaKetchup** [reply]
That's great, and what I was hoping to hear :)

I'll just make a little note in my projects docs and maybe implement a 500MB limit (500MB is an insane number of tiles so is extremely rare, but better safe than sorry I guess).

It looks like Isar is the right database for my package! Just a shame it's not as well 'marketed?' as some of the others (in lists etc).

Thanks for your help!

**Reply by simc** [reply]
Do you have suggestions for better marketing? 🙂

**Reply by JaffaKetchup** [reply]
Not really :D, was only thinking aloud.

Maybe a few blogs (something like Medium)? I noticed there are [already some](https://medium.com/@mustafatahirhussein/exploring-isar-database-for-flutter-f6d45de4acd8), but they don't appear in Google for searches like "Flutter database", only for "Isar Flutter". So maybe a comparison between this and other popular databases? Then you can cheat by using keywords like "SQFlite" and "Moor" :D.

I'm also talking about lists like [this](https://www.dexbytes.com/blog/best-databases-for-flutter-in-2022/) and [this](https://greenrobot.org/news/flutter-databases-a-comprehensive-comparison/): there's not much you can do about these, and I guess recognition will come with time?
Somehow mine managed to get into a [non-robot-created list](https://issues.flutter-croissants.com/editions/3#:~:text=flutter_map_tile_caching%20%3A%20Caching%20maps) (that I've never heard of before) and it's been mentioned in a [Reddit thread](https://www.reddit.com/r/FlutterDev/comments/z1r8ac/make_offline_map_with_flutter_map/) or two - but mine is such as specific niche that marketing it isn't really the point: I just 'advertise' it through the [official flutter_map docs](https://docs.fleaflet.dev/plugins/list) that I wrote. I guess all of these things are out of your control.

(Of course, mentioning your library in any issues/discussions you make also helps make connections *cheeky wink*)

Anyway, I hope this gains more traction!

**Reply by JaffaKetchup** [reply]
When I implement this in my library, I'll make sure to put this one somewhere prominent in my docs :)
